packages/storage-plugin/lib/storage.ts

```diff
@@ -0,0 +1,320 @@
+      source.pipe(destination);
+      try {
+        await new Promise((resolve, reject) => {
+          destination.once('finish', () => resolve(true));
+          source.once('error', reject);
+          destination.once('error', reject);
+        });
+        await this._finalizeItem(opts, timer, fullPath, await fs.hash(fullPath));
+      } catch (e) {
+        await fs.rimraf(fullPath);
+        throw e;
+      }
+    } finally {
+      delete this._inProgressAdditions[name];
+    }
+  }
+
+  private async _addFromWebSocket(opts: ItemOptions, source: WebSocket): Promise<void> {
+    const {name, sha1} = opts;
+    const fullPath = path.join(this._root, toTempName(name));
+    const inProgressInfo: InProgressAddition = {fullPath};
+    this._inProgressAdditions[name] = inProgressInfo;
+    try {
+      const timer = new timing.Timer().start();
+      const destination = fs.createWriteStream(fullPath);
+      const sha1sum = createHash('sha1');
+      let didDigestMatch = false;
+      let recentDigest: string | null = null;
+      try {
+        await new Promise((resolve, reject) => {
+          source.on('message', (data: WebSocket.RawData) => {
+            if (didDigestMatch) {
+              // ignore further chunks if hashes have already matched
+              return;
+            }
+            destination.write(data, (e) => {
+              if (e) {
+                source.close(WS_SERVER_ERROR);
+                reject(e);
+              }
+            });
+            sha1sum.update(data as any);
+            recentDigest = sha1sum.copy().digest('hex');
+            if (_.toLower(recentDigest) === _.toLower(sha1)) {
+              didDigestMatch = true;
+              destination.close(() => resolve(true));
+            }
+          });
+          source.once('close', () => {
+            destination.close(() => resolve(true));
+          });
+          source.once('error', reject);
+          destination.once('error', (e) => {
+            source.close(WS_SERVER_ERROR);
+            reject(e);
+          });
+        });
+        await this._finalizeItem(opts, timer, fullPath, recentDigest ?? sha1sum.digest('hex'));
+      } catch (e) {
+        await fs.rimraf(fullPath);
+        throw e;
+      }
+    } finally {
+      delete this._inProgressAdditions[name];
+    }
+  }
+
+  private async _finalizeItem(
+    opts: ItemOptions,
+    timer: timing.Timer,
+    fullPath: string,
+    actualHashDigest: string,
+  ): Promise<void> {
+    const {name, sha1} = opts;
+    this._log.info(
+      `'${name}' has been added to the server storage within ` +
+      `${timer.getDuration().asMilliSeconds}ms. Verifying hashes.`
+    );
+    if (_.toLower(actualHashDigest) !== _.toLower(sha1)) {
+      throw new StorageArgumentError(
+        `The actual SHA1 hash value '${actualHashDigest}' must be equal ` +
+        `to the expected hash value of '${sha1}' for '${name}'`
+      );
+    }
+    await fs.mv(fullPath, path.join(this._root, name));
```

Comment:
Would it be nice to address that if the same name file already exists, the new one will override the existing one. If the already existing file is dir, an exception would be raised, so this finalizing will fail in this README?

(based on fs.rename in the `mv` behavior)

Reply:
added a comment